### PR TITLE
Update shapelet_transform.py

### PR DESCRIPTION
**Pull request by jmrichardson**
Simple addition to add dtype='object' when saving shapelet

### CODE REVIEW

**Comment by johannfaouzi** [comment]
Ok, so `numba` does not support arrays with `dtype='object'`. I did not keep up to date with their most recent releases, but it seems like they now better support lists than previously, with their own class. I used a tuple to store shapelets, and this approach has its own issues (like the size limit, as you pointed it out in your other issue).

So I think we can address both issues in this PR by using the `numba.typed.List` class to store shapelets. Required changes:

* Between lines 5 and 6: you have to import the `numba.typed.List` class:

```python
from numba.typed import List
```

* Line 532: you need to write it in two lines (otherwise the line is too long):
```python
shapelets = np.asarray(list(chain.from_iterable(shapelets)),
                       dtype='object')
```

* Between lines 578 and 579: you have to add this to convert back the `dtype` from `object` to `float64`:
```python
if shapelets.ndim > 1:
    shapelets = shapelets.astype('float64')
```

* Between lines 605 and 606: you have to store the shapelets in a `numba.typed.List` object:
```python
# Convert the array of arrays of shapelets to a numba.typed.List object
shapelets_list = List()
for x in self.shapelets_:
    shapelets_list.append(x.astype('float64'))
```

* Line 607: you have to use the `shapelets_list` object:
```python
X, window_sizes, shapelets_list, lengths, fit=False)
```

Note: The `LearningShapelets` may have the same issues. To fix the warnings, you just have to add the `dtype='object'`, that is updating line 589 of the `pyts/classification/learning_shapelets.py` file with:
```python
list(chain.from_iterable(self.shapelets_)), dtype='object')
```
There is a similar issue regarding the size of the tuple, but it's less restrictive (it only happens if there are more than 1000 different lengths, because the learned shapelets are stored with 2D-arrays, one for each length). So I think that we can let this code as it is for the moment.

**Comment by jmrichardson** [comment]
Hi, thank you so much for helping me.  I am sorry I don't know numba/numpy enough to be much help.  I did make the updates you mentioned but I am still getting the error message:

The above exception was the direct cause of the following exception:

```
Traceback (most recent call last):
  File "pipeline.py", line 25, in <module>
    X_trn, st = shaplet(X_trn_seg, y_trn_seg, X_trn)  # Shapelet features
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/pyts/transformation/shapelet_transform.py", line 572, in _fit
    for i in range(n_samples))
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/parallel.py", line 1042, in __call__
    self.retrieve()
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/parallel.py", line 921, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 542, in wrap_future_result
    return future.result(timeout=timeout)
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
numba.core.errors.UnsupportedError: Failed in nopython mode pipeline (step: ensure IR is legal prior to lowering)
Tuple 'shapelets' length must be smaller than 1000.
Large tuples lead to the generation of a prohibitively large LLVM IR which causes excessive memory pressure and large compile times.
As an alternative, the use of a 'list' is recommended in place of a 'tuple' as lists do not suffer from this problem.

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/pyts/transformation/shapelet_transform.py", line 72:
@njit()
def _derive_all_squared_distances_fit(
```

File: pyts/transformation/shapelet_transform.py [hunk 1]
```diff
@@ -575,6 +581,8 @@ def _fit(self, X, y, n_shapelets, window_sizes, window_steps,
         # Concatenate the results
         X_dist = np.hstack(X_dist)
         scores = np.concatenate(scores)
+        if shapelets.ndim > 1:
+            shapelets = shapelets.astype('float64')
```
Comment (johannfaouzi) [inline pyts/transformation/shapelet_transform.py:585]: You have to put this after:
```python
shapelets = np.concatenate(shapelets)
```

**Comment by jmrichardson** [comment]
Hi, I noticed that the _fit_one_time_series function also calls _derive_all_distances, so I made similar code updates:

```
    def _fit_one_time_series(
        self, x, X, y, n_timestamps, n_shapelets, window_sizes,
        window_steps, remove_similar, i, rng
    ):
        """Fit one time series."""
        # Extract all shapelets
        shapelets, lengths, start_idx, end_idx = _extract_all_shapelets(
            x, window_sizes, window_steps, n_timestamps)

        shapelets_list = List()
        for x in shapelets:
            shapelets_list.append(x.astype('float64'))

        # Derive distances between shapelets and time series
        X_dist = _derive_all_distances(
            X, window_sizes, shapelets_list, lengths, fit=True)
```

However, for some reason it still doesn't like it:

```
Traceback (most recent call last):
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 431, in _process_worker
    r = call_item()
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/parallel.py", line 253, in __call__
    for func, args, kwargs in self.items]
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/joblib/parallel.py", line 253, in <listcomp>
    for func, args, kwargs in self.items]
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/pyts/transformation/shapelet_transform.py", line 527, in _fit_one_time_series
    X, window_sizes, shapelets_list, lengths, fit=True)
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/pyts/transformation/shapelet_transform.py", line 116, in _derive_all_distances
    X, n_samples, n_timestamps, window_sizes, shapelets, lengths
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/dispatcher.py", line 404, in _compile_for_args
    error_rewrite(e, 'unsupported_error')
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/utils.py", line 80, in reraise
    raise value.with_traceback(tb)
numba.core.errors.UnsupportedError: Failed in nopython mode pipeline (step: ensure IR is legal prior to lowering)
Tuple 'lengths' length must be smaller than 1000.
Large tuples lead to the generation of a prohibitively large LLVM IR which causes excessive memory pressure and large compile times.
As an alternative, the use of a 'list' is recommended in place of a 'tuple' as lists do not suffer from this problem.

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/pyts/transformation/shapelet_transform.py", line 72:
@njit()
def _derive_all_squared_distances_fit(
```

**Review by johannfaouzi — COMMENTED** [review]
Now the error is:
```python
AttributeError: 'tuple' object has no attribute 'ndim'
```
because `shapelets` is a tuple and not an array. You just have to put the two new lines below
```python
shapelets = np.concatenate(shapelets)
```

File: pyts/transformation/shapelet_transform.py [hunk 2]
```diff
@@ -517,9 +518,17 @@ def _fit_one_time_series(
...
+            shapelets_list.append(x.astype('float64'))
+
+        lengths_list = List()
+        for x in lengths:
+            lengths_list.append(x.astype('float64'))
+
         # Derive distances between shapelets and time series
         X_dist = _derive_all_distances(
-            X, window_sizes, shapelets, lengths, fit=True)
+            X, window_sizes, shapelets_list, lengths_list, fit=True)
```
Comment (johannfaouzi) [inline pyts/transformation/shapelet_transform.py:531]: I don't think that it's needed

File: pyts/transformation/shapelet_transform.py [hunk 3]
```diff
@@ -603,8 +615,16 @@ def _transform(self, X):
...
         window_sizes = np.unique(lengths)
 
+        shapelets_list = List()
+        for x in self.shapelets_:
+            shapelets_list.append(x.astype('float64'))
+
+        lengths_list = List()
+        for x in self.shapelets_:
+            lengths_list.append(x.astype('float64'))
+
```
Comment (johannfaouzi) [inline pyts/transformation/shapelet_transform.py:625]: Likewise, `lengths` is not an issue because it's an array with `dtype='int64'`

**Comment by jmrichardson** [comment]
I got this error before:

Tuple '**lengths**' length must be smaller than 1000.

So i was trying to do the same logic lengths tuple as with shapelets tuple.  I am testing the code now on my dataset and seems to be working.  Unfortunately, my test code is on another remote server so I have to code local, commit, pull and then test on remote server.  

Please let me know if you think I should remove the lengths_list code.

I am getting these warnings now though with the latest commit of my fork, but the process is still running:

```
2020-09-12 09:35: Generating shapelet features
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py:66: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function 'impl_append.<locals>.impl'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/listobject.py", line 596:

    def impl(l, item):
    ^

  l.append(item)
/home/jmrichardson/anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/core/ir_utils.py:2031: NumbaPendingDeprecationWarning:
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'item' of function '_append'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "../anaconda3/envs/tipjar/lib/python3.7/site-packages/numba/typed/typedlist.py", line 65:
@njit
def _append(l, item):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
```

**Comment by jmrichardson** [comment]
Line 52 is where i think lengths is converted to tuple

**Comment by jmrichardson** [comment]
Also, if it is easier, I can pickle my X, y dataset and share it with you.  Appreciate all the help

**Comment by johannfaouzi** [comment]
I see the issue, I was not expecting such a large number of shapelets for each time series. No need to provide the data itself, but the shape of `X` could be useful.

Numba didn't like lists at all a while ago (you could create a list inside a function, but you couldn't give a list as a parameter of a function), so that's why there is a lot of conversion between lists and tuples. But now that numba provides their own class for lists, I think that I could clean up the code (and fix all the warnings and errors that you are facing).

**Comment by jmrichardson** [comment]
Shape of dataset:  (36278, 2400)

I just ran the code with smaller dataset shape (313, 20) locally.  I did get the same warning messages but it also failed execution:

```
  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
Traceback (most recent call last):
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\IPython\core\interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-1447c6ec6541>", line 1, in <module>
    X_trn, st = shaplet(X_trn_seg, y_trn_seg, X_trn)  # Shapelet features
  File "D:\Projects\tipjar\tipjar\prepare\shapelet.py", line 15, in shaplet
    features = st.transform(X_trn_seg)
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\pyts\transformation\shapelet_transform.py", line 343, in transform
    return self._transform(X)
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\pyts\transformation\shapelet_transform.py", line 627, in _transform
    X, window_sizes, shapelets_list, lengths_list, fit=False)
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\pyts\transformation\shapelet_transform.py", line 121, in _derive_all_distances
    X, n_samples, n_timestamps, window_sizes, shapelets, lengths
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\numba\core\dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\numba\core\dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "D:\Anaconda3\envs\tipjar\lib\site-packages\numba\core\utils.py", line 80, in reraise
    raise value.with_traceback(tb)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function where at 0x000001FD76FB5438>) with argument(s) of type(s): (bool)
 * parameterized
In definition 0:
    AttributeError: 'Boolean' object has no attribute 'ndim'
    raised from D:\Anaconda3\envs\tipjar\lib\site-packages\numba\core\typing\npydecl.py:1116
In definition 1:
    AttributeError: 'Boolean' object has no attribute 'ndim'
    raised from D:\Anaconda3\envs\tipjar\lib\site-packages\numba\core\typing\npydecl.py:1116
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function where at 0x000001FD76FB5438>)
[2] During: typing of call at D:\Anaconda3\envs\tipjar\lib\site-packages\pyts\transformation\shapelet_transform.py (101)
```

**Comment by johannfaouzi** [comment]
I think that using `numba.typed.List` everywhere is the best guess. I used `List()` instead of `[]` in [_extract_all_shapelets_raw](https://github.com/johannfaouzi/pyts/blob/1aa45589b91a12e8d55db86f1f97dca0b6e99984/pyts/transformation/shapelet_transform.py#L16), renamed this function `_extract_all_shapelets` and removed the old `_extract_all_shapelets` because it was just used to convert lists into tuples, which we don't need anymore.

So, instead of this:
```python
@njit()
def _extract_all_shapelets_raw(x, window_sizes, window_steps, n_timestamps):
    """Extract all the shapelets from a single time series."""
    shapelets = []  # shapelets
    lengths = []  # lengths of shapelets
    start_idx = []  # start index of shapelets (included)
    end_idx = []  # end index of shapelets (excluded)

    for window_size, window_step in zip(window_sizes, window_steps):
        # Derive the new shape and strides
        overlap = window_size - window_step
        shape_new = ((n_timestamps - overlap) // window_step, window_size // 1)
        strides = x.strides[0]
        strides_new = (window_step * strides, strides)

        # Derive strided view of x
        x_strided = as_strided(x, shape=shape_new, strides=strides_new)
        x_strided = np.copy(x_strided)

        # Add shapelets, lengths, start indices and end indices
        shapelets.append(x_strided)
        lengths.append([x_strided.shape[1]] * x_strided.shape[0])
        start_idx.append(np.arange(0, n_timestamps - window_size + 1,
                                   window_step))
        end_idx.append(np.arange(window_size, n_timestamps + 1, window_step))

    return shapelets, lengths, start_idx, end_idx


def _extract_all_shapelets(x, window_sizes, window_steps, n_timestamps):
    """Extract all the shapelets from a single time series."""
    shapelets, lengths, start_idx, end_idx = _extract_all_shapelets_raw(
        x, window_sizes, window_steps, n_timestamps)

    # Convert list to tuple
    shapelets = tuple(shapelets)
    lengths = tuple(lengths)

    return shapelets, lengths, start_idx, end_idx
```
I have this now:
```python
@njit()
def _extract_all_shapelets(x, window_sizes, window_steps, n_timestamps):
    """Extract all the shapelets from a single time series."""
    shapelets = List()  # shapelets
    lengths = List()  # lengths of shapelets
    start_idx = List()  # start index of shapelets (included)
    end_idx = List()  # end index of shapelets (excluded)

    for window_size, window_step in zip(window_sizes, window_steps):
        # Derive the new shape and strides
        overlap = window_size - window_step
        shape_new = ((n_timestamps - overlap) // window_step, window_size // 1)
        strides = x.strides[0]
        strides_new = (window_step * strides, strides)

        # Derive strided view of x
        x_strided = as_strided(x, shape=shape_new, strides=strides_new)
        x_strided = np.copy(x_strided)

        # Add shapelets, lengths, start indices and end indices
        shapelets.append(x_strided)
        lengths.append([x_strided.shape[1]] * x_strided.shape[0])
        start_idx.append(np.arange(0, n_timestamps - window_size + 1,
                                   window_step))
        end_idx.append(np.arange(window_size, n_timestamps + 1, window_step))

    return shapelets, lengths, start_idx, end_idx
```

And I don't get an error anymore regarding `lengths`.

**Comment by johannfaouzi** [comment]
So you don't need this anymore because they are already `numba.typed.List` objects:
https://github.com/jmrichardson/pyts/blob/b48cf2d6f50ee91b75c7340544eb906783132e1b/pyts/transformation/shapelet_transform.py#L521-L527

And this one never was an issue I think:
https://github.com/jmrichardson/pyts/blob/b48cf2d6f50ee91b75c7340544eb906783132e1b/pyts/transformation/shapelet_transform.py#L622-L624

**Comment by johannfaouzi** [comment]
Hey,

Are you still working on shapelets for time series? Did my proposed solution work for you? If so, it would be nice if you could update your code and push it, so that I can merge your PR.

**Pull request closed without merging**